BUG - batch event indexer fails

### Description of the issue

_No response_

### Steps to reproduce the issue

_No response_

### Expected behavior (What should happen)

_No response_

### Actual behavior (What does happen)




2025-09-19T06:21:31.164Z[INFO]	2025-09-19T06:21:31.164Z	5cdb2970-7fbf-4ffc-b692-09b02e681b4d	Received event: {     "version": "0",     "id": "a21b9cc6-b5e9-5f04-046e-dc90fa1627b4",     "detail-type": "Batch Job State Change",     "source": "aws.batch",     "account": "593025701104",     "time": "2025-09-19T06:21:27Z",     "region": "us-west-2",     "resources": [         "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd"     ],     "detail": {         "jobArn": "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd",         "jobName": "glows-l3b-ion-rate-profile-job-1",         "jobId": "b7b5f552-0c07-4bd4-80c1-c3d06422c4dd",         "jobQueue": "arn:aws:batch:us-west-2:593025701104:job-queue/ProcessingJobQueue",         "status": "FAILED",         "attempts": [             {                 "container": {                     "taskArn": "arn:aws:ecs:us-west-2:593025701104:task/AWSBatch-ProcessingComputeEnvironment-spot-72b72e17-1329-3b45-b86a-1e7d669cd39a/0a641d52cf064e8490377a4cd6751e94",                     "logStreamName": "ProcessingJob-glows-l3/default/0a641d52cf064e8490377a4cd6751e94",                     "networkInterfaces": [                         {                             "attachmentId": "f8ab8e57-c52c-485c-9114-29a5ee2c743a",                             "privateIpv4Address": "10.0.179.60"                         }                     ]                 },                 "stoppedAt": 1758262886945,                 "statusReason": "ResourceInitializationError: unable to pull secrets or registry auth: execution resource retrieval failed: unable to retrieve secrets from ssm: service call has been retried 1 time(s): invalid ssm parameters: /imap-sdc/batch-jobs/api-key"             }         ],         "statusReason": "ResourceInitializationError: unable to pull secrets or registry auth: execution resource retrieval failed: unable to retrieve secrets from ssm: service call has been retried 1 time(s): invalid ssm parameters: /imap-sdc/batch-jobs/api-key",         "createdAt": 1758262829264,         "retryStrategy": {             "attempts": 2,             "evaluateOnExit": [                 {                     "onStatusReason": "Your Spot Task was interrupted.",                     "action": "retry"                 },                 {                     "onReason": "*",                     "action": "exit"                 }             ]         },         "stoppedAt": 1758262886945,         "dependsOn": [],         "jobDefinition": "arn:aws:batch:us-west-2:593025701104:job-definition/ProcessingJob-glows-l3:3",         "parameters": {},         "container": {             "image": "593025701104.dkr.ecr.us-west-2.amazonaws.com/glows-l3-repo:latest",             "command": [                 "--instrument",                 "glows",                 "--data-level",                 "l3b",                 "--descriptor",                 "ion-rate-profile",                 "--start-date",                 "20250919",                 "--version",                 "v001",                 "--dependency",                 "imap_glows_l3b_ion-rate-profile-4f53cda1_20250919_v001.json",                 "--upload-to-sdc"             ],             "executionRoleArn": "arn:aws:iam::593025701104:role/SDCStack-ProcessingConstructFargateContainerglowsl3-obv08DMzpbv9",             "volumes": [],             "environment": [                 {                     "name": "IMAP_DATA_ACCESS_URL",                     "value": "https://api.imap-mission.com/api-key"                 }             ],             "mountPoints": [],             "readonlyRootFilesystem": false,             "ulimits": [],             "taskArn": "arn:aws:ecs:us-west-2:593025701104:task/AWSBatch-ProcessingComputeEnvironment-spot-72b72e17-1329-3b45-b86a-1e7d669cd39a/0a641d52cf064e8490377a4cd6751e94",             "logStreamName": "ProcessingJob-glows-l3/default/0a641d52cf064e8490377a4cd6751e94",             "networkInterfaces": [                 {                     "attachmentId": "f8ab8e57-c52c-485c-9114-29a5ee2c743a",                     "privateIpv4Address": "10.0.179.60"                 }             ],             "resourceRequirements": [                 {                     "value": "4096",                     "type": "MEMORY"                 },                 {                     "value": "1",                     "type": "VCPU"                 }             ],             "secrets": [                 {                     "name": "IMAP_API_KEY",                     "valueFrom": "arn:aws:ssm:us-west-2:593025701104:parameter/imap-sdc/batch-jobs/api-key"                 }             ],             "networkConfiguration": {                 "assignPublicIp": "ENABLED"             },             "fargatePlatformConfiguration": {                 "platformVersion": "LATEST"             },             "runtimePlatform": {                 "operatingSystemFamily": "LINUX",                 "cpuArchitecture": "X86_64"             }         },         "tags": {             "resourceArn": "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd"         },         "propagateTags": false,         "platformCapabilities": [             "FARGATE"         ],         "eksAttempts": []     } } | [INFO] 2025-09-19T06:21:31.164Z 5cdb2970-7fbf-4ffc-b692-09b02e681b4d Received event: { "version": "0", "id": "a21b9cc6-b5e9-5f04-046e-dc90fa1627b4", "detail-type": "Batch Job State Change", "source": "aws.batch", "account": "593025701104", "time": "2025-09-19T06:21:27Z", "region": "us-west-2", "resources": [ "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd" ], "detail": { "jobArn": "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd", "jobName": "glows-l3b-ion-rate-profile-job-1", "jobId": "b7b5f552-0c07-4bd4-80c1-c3d06422c4dd", "jobQueue": "arn:aws:batch:us-west-2:593025701104:job-queue/ProcessingJobQueue", "status": "FAILED", "attempts": [ { "container": { "taskArn": "arn:aws:ecs:us-west-2:593025701104:task/AWSBatch-ProcessingComputeEnvironment-spot-72b72e17-1329-3b45-b86a-1e7d669cd39a/0a641d52cf064e8490377a4cd6751e94", "logStreamName": "ProcessingJob-glows-l3/default/0a641d52cf064e8490377a4cd6751e94", "networkInterfaces": [ { "attachmentId": "f8ab8e57-c52c-485c-9114-29a5ee2c743a", "privateIpv4Address": "10.0.179.60" } ] }, "stoppedAt": 1758262886945, "statusReason": "ResourceInitializationError: unable to pull secrets or registry auth: execution resource retrieval failed: unable to retrieve secrets from ssm: service call has been retried 1 time(s): invalid ssm parameters: /imap-sdc/batch-jobs/api-key" } ], "statusReason": "ResourceInitializationError: unable to pull secrets or registry auth: execution resource retrieval failed: unable to retrieve secrets from ssm: service call has been retried 1 time(s): invalid ssm parameters: /imap-sdc/batch-jobs/api-key", "createdAt": 1758262829264, "retryStrategy": { "attempts": 2, "evaluateOnExit": [ { "onStatusReason": "Your Spot Task was interrupted.", "action": "retry" }, { "onReason": "*", "action": "exit" } ] }, "stoppedAt": 1758262886945, "dependsOn": [], "jobDefinition": "arn:aws:batch:us-west-2:593025701104:job-definition/ProcessingJob-glows-l3:3", "parameters": {}, "container": { "image": "593025701104.dkr.ecr.us-west-2.amazonaws.com/glows-l3-repo:latest", "command": [ "--instrument", "glows", "--data-level", "l3b", "--descriptor", "ion-rate-profile", "--start-date", "20250919", "--version", "v001", "--dependency", "imap_glows_l3b_ion-rate-profile-4f53cda1_20250919_v001.json", "--upload-to-sdc" ], "executionRoleArn": "arn:aws:iam::593025701104:role/SDCStack-ProcessingConstructFargateContainerglowsl3-obv08DMzpbv9", "volumes": [], "environment": [ { "name": "IMAP_DATA_ACCESS_URL", "value": "https://api.imap-mission.com/api-key" } ], "mountPoints": [], "readonlyRootFilesystem": false, "ulimits": [], "taskArn": "arn:aws:ecs:us-west-2:593025701104:task/AWSBatch-ProcessingComputeEnvironment-spot-72b72e17-1329-3b45-b86a-1e7d669cd39a/0a641d52cf064e8490377a4cd6751e94", "logStreamName": "ProcessingJob-glows-l3/default/0a641d52cf064e8490377a4cd6751e94", "networkInterfaces": [ { "attachmentId": "f8ab8e57-c52c-485c-9114-29a5ee2c743a", "privateIpv4Address": "10.0.179.60" } ], "resourceRequirements": [ { "value": "4096", "type": "MEMORY" }, { "value": "1", "type": "VCPU" } ], "secrets": [ { "name": "IMAP_API_KEY", "valueFrom": "arn:aws:ssm:us-west-2:593025701104:parameter/imap-sdc/batch-jobs/api-key" } ], "networkConfiguration": { "assignPublicIp": "ENABLED" }, "fargatePlatformConfiguration": { "platformVersion": "LATEST" }, "runtimePlatform": { "operatingSystemFamily": "LINUX", "cpuArchitecture": "X86_64" } }, "tags": { "resourceArn": "arn:aws:batch:us-west-2:593025701104:job/b7b5f552-0c07-4bd4-80c1-c3d06422c4dd" }, "propagateTags": false, "platformCapabilities": [ "FARGATE" ], "eksAttempts": [] } }
  | 2025-09-19T06:21:31.165Z[ERROR] KeyError: 'startedAt'Traceback (most recent call last):  File "/var/task/SDSCode/pipeline_lambdas/indexer.py", line 396, in lambda_handler    return batch_event_handler(event)  File "/var/task/SDSCode/pipeline_lambdas/indexer.py", line 352, in batch_event_handler    started_at_timestamp = event["detail"]["startedAt"] |  

<br class="Apple-interchange-newline">

### Code Snippet:

```python
Code
```

### Additional notes, affected areas, and suggested fixes

_No response_

--- a/sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/indexer.py
+++ b/sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/indexer.py
@@ -348,11 +348,23 @@ def batch_event_handler(event):
     # We injected our table ID into the job name
     job_id = event["detail"]["jobName"].split("-")[-1]
 
-    # Convert startAt and stoppedAt to datetime with timezone
-    started_at_timestamp = event["detail"]["startedAt"]
-    started_at = datetime.fromtimestamp(started_at_timestamp / 1000, tz=timezone.utc)
-    stopped_at_timestamp = event["detail"]["stoppedAt"]
-    stopped_at = datetime.fromtimestamp(stopped_at_timestamp / 1000, tz=timezone.utc)
+    # Convert startedAt and stoppedAt to datetime with timezone
+    # These fields may not always be present, so use .get() and handle None
+    # Default to startedAt and fallback to createdAt if the job never started
+    started_at_timestamp = event["detail"].get(
+        "startedAt", event["detail"].get("createdAt")
+    )
+    stopped_at_timestamp = event["detail"].get("stoppedAt")
+    started_at = (
+        datetime.fromtimestamp(started_at_timestamp / 1000, tz=timezone.utc)
+        if started_at_timestamp is not None
+        else None
+    )
+    stopped_at = (
+        datetime.fromtimestamp(stopped_at_timestamp / 1000, tz=timezone.utc)
+        if stopped_at_timestamp is not None
+        else None
+    )
 
     with db.Session() as session:
         # Get the batch job by its ID

